docs/cookbook/production-deployment.md

```diff
@@ -0,0 +1,308 @@
+  <port protocol=""tcp"" port=""9300""/>
+  <port protocol=""tcp"" port=""9200""/>
+  <destination ipv4=""192.168.0.0/16 ""/>
+</service> 
+EOT
+firewall-cmd --permanent --zone=public --add-service=elasticsearch
+firewall-cmd --reload
+```
+We will also make elasticsearch server listen on 192.168.0.0 subnet by modifying `/etc/elasticsearch/elasticsearch.yml`.
+```
+network.host: 0.0.0.0
+service elasticsearch restart
+```
+
+### Docker Image Building
+
+We want to deploy changes that we made in git repository to production server. We deploy the `php-fpm` container that encapsulates `project-base` repository.
+First we need to clone `project-base` repository with the specific tag or commit hash or HEAD.
+```
+git clone https://github.com/shopsys/project-base.git
+git checkout $commit_hash
+```
+Then we setup environment for building the image with the correct data for production server.
+Now we create configuration file for domains.
+```
+echo $'domains:
+    -   id: 1
+        name: <YOUR_DOMAIN_NAME_HERE>
+        locale: en
+' > app/config/domains.yml
+```
+For each domain we need to create config with domain url.
+```
+echo $'domains_urls:
+    -   id: 1
+        url: https://<YOUR_DOMAIN_HERE>
+' >  app/config/domains_urls.yml
+```
+We remove `.git` folder so the built image doesn't have additional data.
+After the `project-base` is setup correctly, we launch the build of php-fpm container by docker build command.
+```
+docker build \
+    -f ./docker/php-fpm/Dockerfile-production \
+    -t production-php-fpm \
+    --compress \
+    --build-arg www_data_uid=1000 --build-arg www_data_gid=1000 \
+    --build-arg github_oauth_token=PERSONAL_ACCESS_TOKEN_FROM_GITHUB \
+    .
+```
+Replace the `PERSONAL_ACCESS_TOKEN_FROM_GITHUB` string by the token generated on [Github -> Settings -> Developer Settings -> Personal access tokens](https://github.com/settings/tokens/new?scopes=repo&description=Composer+API+token).
+There are also settings for user id and group id under who the ownership of files will be because php-fpm execution needs workers that are not allowed to runs under root user group.
+With `f` parameter we set path to Dockerfile that builds image.
+With `t` parameter we set the name of built image.
+
+If we are building the image on different server than production server, we can push built image into docker repository of production server via ssh.
+We use `-oStrictHostKeyChecking=no` argument to have ssh connection without the prompt that asks about adding target server record into `known_hosts` ssh configuration.
+We also want to establish connection to the server without prompting for password so we will use [key exchange method](http://sshkeychain.sourceforge.net/mirrors/SSH-with-Keys-HOWTO/SSH-with-Keys-HOWTO-4.html).
+```
+docker save production-php-fpm | gzip | ssh -oStrictHostKeyChecking=no -i <PRIVATE_KEY_PATH> root@<YOUR_DOMAIN_HERE> 'gunzip | docker load'
```

Comment:
Why you just dont push image into docker hub with some tag and pull it from it? you wouldnt have to commit it then and just overwrite docker-compose using `yq write`.

Reply:
i am not sure whether this is secured. If users would use some private registry...